frame/election-providers/src/two_phase/mod.rs

```diff
@@ -0,0 +1,1622 @@
+		<Snapshot<T>>::put(RoundSnapshot { voters, targets });
+	}
+
+	/// Build the witness data from the snapshot metadata, if it exists. Else, returns `None`.
+	fn build_witness() -> Option<WitnessData> {
+		let metadata = Self::snapshot_metadata()?;
+		Some(WitnessData {
+			voters: metadata.voters_len as u32,
+			targets: metadata.targets_len as u32,
+		})
+	}
+
+	/// Checks the feasibility of a solution.
+	///
+	/// This checks the solution for the following:
+	///
+	/// 0. **all** of the used indices must be correct.
+	/// 1. present correct number of winners.
+	/// 2. any assignment is checked to match with [Snapshot::voters].
+	/// 3. for each assignment, the check of `ElectionDataProvider` is also examined.
+	/// 4. the claimed score is valid.
+	fn feasibility_check(
+		solution: RawSolution<CompactOf<T>>,
+		compute: ElectionCompute,
+	) -> Result<ReadySolution<T::AccountId>, FeasibilityError> {
+		let RawSolution {
+			compact,
+			score,
+			round,
+		} = solution;
+
+		// first, check round.
+		ensure!(Self::round() == round, FeasibilityError::InvalidRound);
+
+		// winners are not directly encoded in the solution.
+		let winners = compact.unique_targets();
+
+		// TODO: this can probably checked ahead of time, both in signed and unsigned story.
+		let desired_targets =
+			Self::desired_targets().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		ensure!(
+			winners.len() as u32 == desired_targets,
+			FeasibilityError::WrongWinnerCount,
+		);
+
+		// read the entire snapshot.
+		let RoundSnapshot {
+			voters: snapshot_voters,
+			targets: snapshot_targets,
+		} = Self::snapshot().ok_or(FeasibilityError::SnapshotUnavailable)?;
+
+		// ----- Start building. First, we need some closures.
+		let voter_at = crate::voter_at_fn!(snapshot_voters, T::AccountId, T);
+		let target_at = crate::target_at_fn!(snapshot_targets, T::AccountId, T);
+
+		// first, make sure that all the winners are sane.
+		let winners = winners
+			.into_iter()
+			.map(|i| target_at(i).ok_or(FeasibilityError::InvalidWinner))
+			.collect::<Result<Vec<T::AccountId>, FeasibilityError>>()?;
+
+		// Then convert compact -> Assignment. This will fail if any of the indices are gibberish.
+		// that winner indices are already checked.
+		let assignments = compact
+			.into_assignment(voter_at, target_at)
+			.map_err::<FeasibilityError, _>(Into::into)?;
+
+		// Ensure that assignments is correct.
+		let _ = assignments
+			.iter()
+			.map(|ref assignment| {
+				// check that assignment.who is actually a voter (defensive-only).
+				// NOTE: This check is fully defensive as we know that this assignment.who is
+				// a voter, what we don't know is stake and targets. Essentially, we have lost the
+				// index after converting `compact` -> `assignment`. This can be optimized in the
+				// future.
+				let (_voter, _stake, targets) = snapshot_voters
+					.iter()
+					.find(|(v, _, _)| v == &assignment.who)
+					.ok_or(FeasibilityError::InvalidVoter)?;
```

Comment:
I'm curious about the performance cost of this.

Reply:
Yes feasibility check is quite unoptimized in general, rather tuned more for veracity.

Reply:
I did a quitck experiment with this diff:
```diff
diff --git a/frame/election-providers/src/two_phase/mod.rs b/frame/election-providers/src/two_phase/mod.rs
index 3f7b8a58a4..9988901fae 100644
--- a/frame/election-providers/src/two_phase/mod.rs
+++ b/frame/election-providers/src/two_phase/mod.rs
@@ -1064,6 +1064,13 @@ where
                let voter_at = crate::voter_at_fn!(snapshot_voters, T::AccountId, T);
                let target_at = crate::target_at_fn!(snapshot_targets, T::AccountId, T);
 
+               use sp_std::collections::btree_map::BTreeMap;
+               let mut voter_indices: BTreeMap<T::AccountId, usize> = BTreeMap::new();
+               for (i, v) in snapshot_voters.iter().enumerate()  {
+                       voter_indices.insert(v.0.clone(), i);
+               }
+
+
                // first, make sure that all the winners are sane.
                let winners = winners
                        .into_iter()
@@ -1085,10 +1092,11 @@ where
                                // a voter, what we don't know is stake and targets. Essentially, we have lost the
                                // index after converting `compact` -> `assignment`. This can be optimized in the
                                // future.
-                               let (_voter, _stake, targets) = snapshot_voters
-                                       .iter()
-                                       .find(|(v, _, _)| v == &assignment.who)
-                                       .ok_or(FeasibilityError::InvalidVoter)?;
+
+                               let voter_index = voter_indices.get(&assignment.who)
+                                       .ok_or(FeasibilityError::InvalidVoter)?.clone();
+                               let (_voter, _stake, targets) = &snapshot_voters[voter_index];
+
                                // check that all of the targets are valid based on the snapshot.
                                if assignment
                                        .distribution
@@ -1101,8 +1109,9 @@ where
                        })
                        .collect::<Result<(), FeasibilityError>>()?;
 
-               // ----- Start building support. First, we need some more closures.
-               let stake_of = stake_of_fn!(snapshot_voters, T::AccountId);
+               let stake_of = |who: &T::AccountId| -> crate::VoteWeight {
+                       snapshot_voters[voter_indices.get(&who).cloned().unwrap_or_default()].1
+               };
 
                // This might fail if the normalization fails. Very unlikely. See `integrity_test`.
                let staked_assignments = assignment_ratio_to_staked_normalized(assignments, stake_of)

```

here the result with btree:
```
/// Weight functions for pallet_two_phase_election_provider.
pub struct WeightInfo<T>(PhantomData<T>);
impl<T: frame_system::Config> pallet_two_phase_election_provider::WeightInfo for WeightInfo<T> {
	fn feasibility_check(v: u32, t: u32, a: u32, d: u32, ) -> Weight {
		(0 as Weight)
			// Standard Error: 50_000
			.saturating_add((5_517_000 as Weight).saturating_mul(v as Weight))
			// Standard Error: 79_000
			.saturating_add((1_143_000 as Weight).saturating_mul(t as Weight))
			// Standard Error: 79_000
			.saturating_add((14_239_000 as Weight).saturating_mul(a as Weight))
			// Standard Error: 195_000
			.saturating_add((1_910_000 as Weight).saturating_mul(d as Weight))
			.saturating_add(T::DbWeight::get().reads(3 as Weight))
	}
}
```
Here the result without btree:
```
/// Weight functions for pallet_two_phase_election_provider.
pub struct WeightInfo<T>(PhantomData<T>);
impl<T: frame_system::Config> pallet_two_phase_election_provider::WeightInfo for WeightInfo<T> {
	fn feasibility_check(v: u32, _t: u32, a: u32, d: u32, ) -> Weight {
		(0 as Weight)
			// Standard Error: 54_000
			.saturating_add((4_232_000 as Weight).saturating_mul(v as Weight))
			// Standard Error: 86_000
			.saturating_add((15_901_000 as Weight).saturating_mul(a as Weight))
			// Standard Error: 212_000
			.saturating_add((4_464_000 as Weight).saturating_mul(d as Weight))
			.saturating_add(T::DbWeight::get().reads(3 as Weight))
	}
}
```

Reply:
While we're at it, I would add that in a correct solution, we want to fetch the stake of each voter exactly once. So maybe the cost of building a btree-map once and reading each key only once for stake is less worth it. Instead, I think if we re-use the same in the other loops as well, it should get even better.

Reply:
By curiosity I tried to do this [diff.txt](https://github.com/paritytech/substrate/files/5789289/diff.txt)
And benchmark but then result seemed a bit weird so I didn't push further.

Anyway this shouldn't be consider a blocker for this PR, but yes would be great if we can find some improvment in this function is it is one bottleneck.

Reply:
The dependency on `a` reduced from ~30 when using linear search to ~12 when using a btree cache on my machine when doing wasm benchmarks. Both this part, and the stake_of_fn that are used in feasibility-check which is onchain are somewhat optimized now.

Reply:
see: https://github.com/paritytech/substrate/pull/7319/commits/000ecdc89b512ed83a18573a99de65963888afd4

benchmark before optimization: https://github.com/paritytech/substrate/pull/7319/commits/e9f4437adf2aba11ddd6747eec155f27aa23aed5

benchmark after optimization: [running]